docs: add doc about running docs locally (#36379)
angular/angular: aio/content/guide/docs-running-locally.md

For more detail, see [Checking your remote URLs](guide/docs-running-locally#check-remotes).

</div>

## How to squash

Squash when you have multiple commits but want all the work in one commit.

1. On your feature branch: `git rebase -i HEAD~#` where `#` is how many commits you want to squash.
For example, if you have 3 commits to squash into one, the command would be `git rebase -i HEAD~3`.
1. Your default editor opens with a list of your commits at the top.
`Pick` the one at the top and `fixup` (`f`) the rest in the list.
1. Save and exit your editor.
To double check you've done it correctly, do a `git log` or `git log --oneline` to see the commit log.
Your fixed up ones should be no longer in the list and the one you picked should be the only one left of your commits on the feature branch.

## How to amend a commit

Amend a commit when you want to change the commit message.

1. On your **feature branch**: `git commit --amend`
1. Your default editor opens.
  Change your commit message.
  Save and close.
1. Back at the command line, run `git log` or `git log --oneline` to confirm that your commit message is as you intend.

> 1. Push up to GitHub.

Reviewer: Maybe add `git push` again?

Author: Thanks for your suggestions, @BioPhoton. I've added them all. 🌟

# Running the Angular docs locally

When contributing to the angular docs, it is helpful to run the docs locally to check your changes as you work.
This guide walks you through the steps of setting up the [Angular repo](https://github.com/angular/angular) locally for the first time.

## Prerequisites

* You need a [GitHub](https://github.com/join) account and a basic [knowledge of git](https://guides.github.com/introduction/git-handbook/).
* Make sure you've read the Angular.io [CONTRIBUTING.md](https://github.com/angular/angular/blob/master/CONTRIBUTING.md).
* Familiarity with the [Angular documentation style guide](guide/docs-style-guide).

<hr />

## Getting your own copy of the repo

The following sections show you how to fork the Angular repo in GitHub, get a local copy of you can work on, and how to run the docs locally.

{@a github}

### In GitHub

The two preliminary steps in this section are in GitHub.
After completing these two steps, continue on to the following section which shows you how to use the terminal to complete your local Angular setup.

> 1. [Fork](https://help.github.com/en/github/getting-started-with-github/fork-a-repo) [angular/angular](https://github.com/angular/angular).

Author: Oh that's good. Thanks!

<code-example language="sh" class="code-shell">

git push --set-upstream origin your-branch

</code-example>

{@a check-remotes}

### Checking your remote URLs

These instructions assume that your git remotes are as follows:

  * `origin`: points to your fork on GitHub
  * `upstream`: points to the `angular/angular` repo

To confirm your remotes, run the following at the command line.

<code-example language="sh" class="code-shell">

git remote --v

</code-example>

The output should be as follows:

<code-example language="sh" class="code-shell">

  origin	https://github.com/your-github-name/angular.git (fetch)
  origin	https://github.com/your-github-name/angular.git (push)
  upstream	https://github.com/angular/angular.git (fetch)
  upstream	https://github.com/angular/angular.git (push)

</code-example>

</div>

{@a fixes-issue}

## Referencing an issue in your commit message

If you're fixing an issue, add the issue number preceded by `Fixes` so GitHub closes the issue automatically on merge as follows:

<code-example language="sh" class="code-shell">

 docs: fix typo in template syntax

 Fixes ####

</code-example>

Replace `####` with the issue number you're fixing.

> Alternatively, you can use the word `Closes`.

Author: Linked instead to github docs.Thanks!

1. Save and exit your editor.
To double check you've done it correctly, do a `git log` or `git log --oneline` to see the commit log.
Your fixed up ones should be no longer in the list and the one you picked should be the only one left of your commits on the feature branch.

## How to amend a commit

Amend a commit when you want to change the commit message.

1. On your **feature branch**: `git commit --amend`
1. Your default editor opens.
  Change your commit message.
  Save and close.
1. Back at the command line, run `git log` or `git log --oneline` to confirm that your commit message is as you intend.
1. Push up to GitHub with `git push`.
  Your PR should reflect the updated commit message.
  You might need `-f` to force the push.
  When force pushing, be sure you understand what it entails before you do it.
  If you're collaborating with someone else, coordinate any force pushing as it overwrites anything else on that branch in GitHub.

## Clearing out commits that aren't yours in your PR

Sometimes when you merge, as in `git pull`,  `master` into your feature branch, you can find a number of commits that aren't yours in your PR.

> To avoid this, consider using `rebase` instead of `merge`.

Author: > Why would you "git pull master into your feature branch"?

You wouldn't, but I think this is how many people end up with those extraneous commits in their PR's. You make a good point though, we shouldn't put how to make the mistake in the documentation, just the fix. I'll remove it.

Before doing this, since your branch is in an undesirable state, back up your work.

> Copy and paste it somewhere outside of your project so that when `git` does its work it doesn't overwrite your back up.

Author: Yes, that was my thinking too. If someone is unsure when trying to fix their git problem, knowing that they have a back up is reassuring.

# Running the Angular docs locally

When contributing to the Angular docs, it is helpful to run the docs locally to check your changes as you work.
This guide walks you through the steps of setting up the [Angular repo](https://github.com/angular/angular) locally for the first time.

## Prerequisites

* You need a [GitHub](https://github.com/join) account and a basic [knowledge of git](https://guides.github.com/introduction/git-handbook/).
  If you use Windows, you may find [Git for Windows](https://gitforwindows.org/)
 helpful.
* Make sure you've read the Angular.io [CONTRIBUTING.md](https://github.com/angular/angular/blob/master/CONTRIBUTING.md).
* Familiarity with the [Angular documentation style guide](guide/docs-style-guide).

> <hr />

Reviewer 2: We can remove the <hr /> tags. I think the latest styles update makes them unnecessary?

> ## Getting your own copy of the repo

Reviewer 2: In our new contributor's guide, we point users to the CONTRIBUTING.md file for this information. Do you think that's sufficient?

If you think this content is important to keep, let's create a separate topic for this information. That way, we can link to it from other topics as well!

The following sections show you how to fork the Angular repo in GitHub, get a local copy  you can work on, and how to build and run the docs app locally.

{@a github}

> ### In GitHub

Reviewer 2: ```suggestion
### Forking a repository using the GitHub UI
```

The two preliminary steps in this section are in GitHub.
After completing these two steps, continue on to the following section which shows you how to use the terminal to complete your local Angular setup.

1. [Fork](https://help.github.com/en/github/getting-started-with-github/fork-a-repo) the[angular/angular](https://github.com/angular/angular).
1. In your new fork, click the "Clone or Download" button to expand the Clone functionality and click the copy-to-clipboard icon to copy the URL of your fork.

<div class="lightbox">
  <img src="generated/images/guide/docs-running-locally/how-to-clone-repo.png"
    alt="screenshot of clone button on GitHub">
</div>

> ### In terminal

Reviewer 2: ```suggestion
### Building the documentation locally
```

> Switch to your terminal and take the following steps:

Reviewer 2: ```suggestion
After you've created a local copy of your forked repository, you can build the Angular documentation. 
```

> 1. Clone your fork: `git clone the-angular-repo-you-want-to-clone`, where `the-angular-repo-you-want-to-clone` is the link you copied in GitHub's ["Clone or Download" dialogue](guide/docs-running-locally#github).

Reviewer 2: Line 39 isn't necessary.

> 1. Go into the `angular` directory: `cd angular`

Reviewer 2: ```suggestion
1. Navigate to the `angular` directory: `cd angular`
```

1. Checkout a new branch where you'll do your work: `git checkout -b yourname-branch-topic`

> 1. Go into the `aio` directory: `cd aio`

Reviewer 2: ```suggestion
1. Navigate to the `aio` directory: `cd aio`
```

1. Setup `aio` with `yarn`: `yarn setup`
1. Start the app and watch for changes: `yarn serve-and-sync`
1. Open a second tab or window so you can issue git commands as you work.

<div class="alert is-helpful">

> As an alternative to running `yarn serve-and-sync`, you can instead run `yarn start` and `yarn docs-watch` in two separate tabs.

Reviewer 2: ```suggestion
As an alternative to running `yarn serve-and-sync`, you can instead run `yarn start` and `yarn docs-watch` in two separate terminals.
```

</div>

> Note that if you already have yarn running and you switch branches, the yarn processes might break.

Reviewer 2: ```suggestion
If you already have yarn running and you switch branches, the yarn processes might break.
```

Should this be a warning?

In this case, `Ctrl`/`Cmd`+`C` and rerun the `yarn` command(s).

> ### In the browser

Reviewer 2: ```suggestion
### Viewing the documentation
```

If the above steps were successful, you should be able to view your local copy of the docs in your browser.

> 1. Go to [localhost:4200](http://localhost:4200).

Reviewer 2: A single numbered list seems odd. Maybe just:

To view the documentation, open a browser window and navigate to localhost:4200.

> ## Where the docs are

Reviewer 2: ```suggestion
## Finding the source files for documentation
```

> Many of the docs on [angular.io](https://www.angular.io) have an _Edit_, and often _View source_, icon at the top right, which opens the file on GitHub.

Reviewer 2: ```suggestion
Sometimes, it's difficult to find the source file of the topic you want to update. Many of the docs on [angular.io](https://www.angular.io) have an _Edit_, and often _View source_, icon at the top right, which opens the file on GitHub.
```

The path in the repo on GitHub is the same as the path in your local clone and shows you where that file resides.

Groups of docs&mdash;such as the guides or example apps&mdash;share particular locations within the Angular repo.
The following table shows the location of docs by type.

<table>

  <tr>
    <th>Type of doc</th>
    <th>Location</th>
  </tr>
  <tr>
    <td>Guides</td>
    <td><code>aio/content/guide</code></td>
  </tr>
  <tr>
    <td>Getting Started</td>
    <td><code>aio/content/start</code></td>
  </tr>
  <tr>
    <td>Tutorial (Tour of Heroes)</td>
    <td><code>aio/content/tutorial</code></td>
  </tr>
  <tr>
    <td>Example apps</td>
    <td><code>aio/content/examples</code></td>
  </tr>
  <tr>
    <td>API docs</td>
    <td>Go to the API doc in question on angular.io and look near the top in the table to find the source. That's the file you edit in the API docs.</td>
  </tr>

</table>

> ## Getting your work on GitHub

Reviewer 2: Maybe just point to CONTRIBUTING.md?
